raiden/utils/migrations/v18_to_v19.py

```diff
@@ -0,0 +1,162 @@
+            )
+            if not affected_state_change:
+                continue
+
+            assert 'block_hash' not in data, 'v18 state changes cant contain blockhash'
+            block_number = int(data['block_number'])
+            data['block_hash'] = cache.get(block_number)
+
+            updated_state_changes.append(StateChangeRecord(
+                state_change_identifier=state_change.state_change_identifier,
+                data=json.dumps(data),
+            ))
+
+        storage.update_state_changes(updated_state_changes)
+
+
+def _add_blockhash_to_events(storage: SQLiteStorage, cache: BlockHashCache) -> None:
+    """Adds blockhash to all ContractSendXXX events"""
+    for events_batch in storage.batch_query_event_records(batch_size=500):
+        updated_events = []
+        for event in events_batch:
+            data = json.loads(event.data)
+            if 'raiden.transfer.events.ContractSend' not in data['_type']:
+                continue
+
+            assert 'triggered_by_block_hash' not in data, 'v18 events cant contain blockhash'
+            # Get the state_change that triggered the event and if it has
+            # a block number get its hash. If not fall back to latest.
+            matched_state_changes = storage.get_statechanges_by_identifier(
+                from_identifier=event.state_change_identifier,
+                to_identifier=event.state_change_identifier,
+            )
+            result_length = len(matched_state_changes)
+            msg = 'multiple state changes should not exist for the same identifier'
+            assert result_length == 1, msg
+
+            statechange_data = json.loads(matched_state_changes[0])
+            if 'block_hash' in statechange_data:
+                block_hash = statechange_data['block_hash']
+            elif 'block_number' in statechange_data:
+                block_number = int(statechange_data['block_number'])
+                block_hash = cache.get(block_number)
+            data['triggered_by_block_hash'] = block_hash
+
+            updated_events.append(EventRecord(
+                event_identifier=event.event_identifier,
+                state_change_identifier=event.state_change_identifier,
+                data=json.dumps(data),
+            ))
+
+        storage.update_events(updated_events)
+
+
+def _transform_snapshot(
+        raw_snapshot: Dict[Any, Any],
+        storage: SQLiteStorage,
+        cache: BlockHashCache,
+) -> str:
+    """Upgrades a single snapshot by adding the blockhash to it and to any pending transactions"""
+    snapshot = json.loads(raw_snapshot)
+    block_number = int(snapshot['block_number'])
+    snapshot['block_hash'] = cache.get(block_number)
+
+    pending_transactions = snapshot['pending_transactions']
+    new_pending_transactions = []
+    for transaction_data in pending_transactions:
+        if 'raiden.transfer.events.ContractSend' not in transaction_data['_type']:
```

Comment:
This should be an assert and not a test. `pending_transactions` should **only** have transactions, and all transactions should be `ContractSend*`

Reply:
Good point. Done in https://github.com/raiden-network/raiden/pull/3550/commits/6b0af26081b71b1f230570ddf3b8b2bd804fb5ac